File: clients/python/src/volant/assignor.py
"""Kafka-style range partition assignor (matches volant_broker::range_assign)."""

from __future__ import annotations


def range_assign(num_partitions: int, member_ids: list[str]) -> list[list[int]]:
    """Assign ``num_partitions`` to ``member_ids`` using the range algorithm.

    Members are sorted by id. For ``n`` partitions and ``m`` members:
    ``base = n // m``, ``extra = n % m``; sorted member ``i`` gets
    ``base + (1 if i < extra else 0)`` consecutive partitions.

    Returns a parallel list of partition lists in the original member order.
    Empty members or ``num_partitions <= 0`` yields an empty list per member.
    """
    ids = list(member_ids)
    m = len(ids)
    if m == 0 or num_partitions <= 0:
        return [[] for _ in range(m)]

    indexed = list(enumerate(ids))
    indexed.sort(key=lambda item: item[1])

    base = num_partitions // m
    extra = num_partitions % m

    result: list[list[int]] = [[] for _ in range(m)]
    nxt = 0
    for rank, (orig_idx, _) in enumerate(indexed):
        count = base + (1 if rank < extra else 0)
        result[orig_idx] = list(range(nxt, nxt + count))
        nxt += count
    return result


def range_assign_multi(
    member_ids: list[str],
    member_topics: list[list[str]],
    partition_counts: dict[str, int],
) -> list[list[tuple[str, int]]]:
    """Range-assign each topic independently to members subscribed to it.

    ``member_topics[i]`` is the subscription list for ``member_ids[i]``.
    ``partition_counts`` maps topic name → partition count. Topics missing
    from ``partition_counts`` are skipped. Returns ``assignments[i]`` as
    ``(topic, partition)`` pairs for member ``i``, sorted by topic then
    partition.
    """
    ids = list(member_ids)
    if len(ids) != len(member_topics):
        raise ValueError("member_ids and member_topics must have the same length")
    m = len(ids)
    out: list[list[tuple[str, int]]] = [[] for _ in range(m)]
    if m == 0:
        return out

    topics: list[str] = []
    for subscribed in member_topics:
        topics.extend(subscribed)
    topics.sort()
    deduped: list[str] = []
    for topic in topics:
        if not deduped or deduped[-1] != topic:
            deduped.append(topic)
    topics = deduped

    counts = partition_counts if partition_counts is not None else {}
    for topic in topics:
        n = counts.get(topic)
        if n is None:
            continue
        sub_ids: list[str] = []
        sub_orig: list[int] = []
        for i, subscribed in enumerate(member_topics):
            if any(t == topic for t in subscribed):
                sub_ids.append(ids[i])
                sub_orig.append(i)
        if not sub_ids:
            continue
        parts = range_assign(int(n), sub_ids)
        for j, ps in enumerate(parts):
            orig = sub_orig[j]
            for p in ps:
                out[orig].append((topic, int(p)))

    for assignment in out:
        assignment.sort()
    return out

File: clients/go/assignor.go
package volant

import "sort"

// RangeAssign assigns numPartitions to memberIDs using the Kafka-style
// range algorithm (same as volant_broker::range_assign).
//
// Members are sorted by id. For n partitions and m members:
// base = n/m, extra = n%m; sorted member i gets base+(i<extra ? 1 : 0)
// consecutive partitions. The returned slice is parallel to memberIDs
// (original order). Empty members or 0 partitions yields an empty list
// per member.
func RangeAssign(numPartitions uint32, memberIDs []string) [][]uint32 {
	m := len(memberIDs)
	result := make([][]uint32, m)
	for i := range result {
		result[i] = []uint32{}
	}
	if m == 0 || numPartitions == 0 {
		return result
	}

	type pair struct {
		orig int
		id   string
	}
	indexed := make([]pair, m)
	for i, id := range memberIDs {
		indexed[i] = pair{orig: i, id: id}
	}
	sort.SliceStable(indexed, func(i, j int) bool {
		return indexed[i].id < indexed[j].id
	})

	base := numPartitions / uint32(m)
	extra := numPartitions % uint32(m)
	var next uint32
	for rank, p := range indexed {
		count := base
		if uint32(rank) < extra {
			count++
		}
		parts := make([]uint32, count)
		for i := uint32(0); i < count; i++ {
			parts[i] = next + i
		}
		next += count
		result[p.orig] = parts
	}
	return result
}

// RangeAssignMulti range-assigns each topic independently to members
// subscribed to that topic (same as volant_broker::range_assign_multi).
//
// memberTopics[i] is the subscription list for memberIDs[i].
// partitionCounts maps topic → partition count. Topics missing from
// the map are skipped. Returns assignments[i] as (topic, partition)
// pairs for member i, sorted by topic then partition.
func RangeAssignMulti(memberIDs []string, memberTopics [][]string, partitionCounts map[string]uint32) [][]Assignment {
	if len(memberIDs) != len(memberTopics) {
		panic("RangeAssignMulti: memberIDs and memberTopics must have the same length")
	}
	m := len(memberIDs)
	out := make([][]Assignment, m)
	for i := range out {
		out[i] = []Assignment{}
	}
	if m == 0 {
		return out
	}

	var topics []string
	for _, subscribed := range memberTopics {
		topics = append(topics, subscribed...)
	}
	sort.Strings(topics)
	deduped := topics[:0]
	for _, topic := range topics {
		if len(deduped) == 0 || deduped[len(deduped)-1] != topic {
			deduped = append(deduped, topic)
		}
	}
	topics = deduped

	for _, topic := range topics {
		n, ok := partitionCounts[topic]
		if !ok {
			continue
		}
		var subIDs []string
		var subOrig []int
		for i, subscribed := range memberTopics {
			for _, t := range subscribed {
				if t == topic {
					subIDs = append(subIDs, memberIDs[i])
					subOrig = append(subOrig, i)
					break
				}
			}
		}
		if len(subIDs) == 0 {
			continue
		}
		parts := RangeAssign(n, subIDs)
		for j, ps := range parts {
			orig := subOrig[j]
			for _, p := range ps {
				out[orig] = append(out[orig], Assignment{Topic: topic, Partition: p})
			}
		}
	}

	for i := range out {
		sort.Slice(out[i], func(a, b int) bool {
			if out[i][a].Topic != out[i][b].Topic {
				return out[i][a].Topic < out[i][b].Topic
			}
			return out[i][a].Partition < out[i][b].Partition
		})
	}
	return out
}
